modules/porous_flow/include/userobjects/AdvectiveFluxCalculator.h

```diff
@@ -0,0 +1,116 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#ifndef ADVECTIVEFLUXCALCULATOR_H
+#define ADVECTIVEFLUXCALCULATOR_H
+
+#include "ElementLoopUserObject.h"
+
+/**
+ * Computes K_ij and R+ and R- detailed in
+ * D Kuzmin and S Turek "High-resolution FEM-TVD schemes based on a fully multidimensional flux
+ * limiter" Journal of Computational Physics 198 (2004) 131-158
+ *
+ * K_ij is a measure of flux from node i to node j
+ * R^+_i and R^-_i quantify how much antidiffusion to allow around node i
+ **/
+class AdvectiveFluxCalculator;
+
+template <>
+InputParameters validParams<AdvectiveFluxCalculator>();
+
+class AdvectiveFluxCalculator : public ElementLoopUserObject
```

Comment:
Use `ElementUserObject` if possible. Not sure you have any limitations that would prevent it. From looking at your code briefly, it should be doable.

Reply:
You should also be able to compute the edge valency in this user object and then you would pull it out into the kernel via your defined API.

Reply:
I'm not sure why you're proposing `ElementUserObject`, @andrsd .  I want to do a loop over elements, computing my `_kij` and your `ElementLoopUserObject`, with its `pre()` and `computeElement()` methods seem to be exactly what i want.

Reply:
`ElementUserObject::execute()` is called on every element, that is equivalent to `ElementLoopUserObject::computeElement()` except all reinits are done for you so you do not have to deal with them. The `pre()` method would be equivalent to `ElementUserObject::initialize()`.

The difference between the 2 is that all `ElementUserObjects` are executed in one mesh sweep. While `ElementLoopUserObjects` will do 1 mesh sweep per instance.